File: v2/realsense/main.py
import serial
import time
import os
import numpy as np
import plotly.graph_objs as go
import utm
import pyrealsense2 as rs
from scipy.spatial.transform import Rotation
import threading
from queue import Queue
import logging
import queue

GPS_PORT = 'COM4'
GPS_BAUD_RATE = 57600
SENSOR_HEIGHT = 0.973  # meters
SENSOR_TILT = 0  # Degrees
ANGLE_FROM_GPS = 0  # Degrees
DISTANCE_FROM_GPS = 0  # meters
SENSOR_ORIENTATION = 0  # Degrees
MAX_DISTANCE_FROM_SENSOR = 4  # meters

logging.basicConfig(level=logging.DEBUG, format='%(asctime)s - %(levelname)s - %(message)s')

class DataRecorder:
    def __init__(self):
        self.start_time = time.time()
        self.session_folder = f"data_{time.strftime('%Y%m%d_%H%M%S')}"
        os.makedirs(self.session_folder, exist_ok=True)
        self.gps_file = open(os.path.join(self.session_folder, 'gps_data.txt'), 'w')
        self.pointcloud_file = open(os.path.join(self.session_folder, 'pointcloud_data.npy'), 'wb')
        self.processed_file = open(os.path.join(self.session_folder, 'processed_data.txt'), 'w')
        self.performance_file = open(os.path.join(self.session_folder, 'performance_data.txt'), 'w')
        
        self.latest_pointcloud = None
        self.latest_gps = None
        self.current_heading = 0
        self.pointcloud_lock = threading.Lock()
        self.gps_lock = threading.Lock()
        self.processing_queue = Queue()
        
        self.gps_thread = threading.Thread(target=self.gps_loop)
        self.realsense_thread = threading.Thread(target=self.realsense_loop)
        self.pipeline = None
        self.align = None
        self.pipeline_started = False
        self.pipeline_lock = threading.Lock()
        self.processing_thread = threading.Thread(target=self.processing_loop)
        
        self.stop_event = threading.Event()

        # Performance monitoring
        self.total_processing_time = 0
        self.total_processed_frames = 0
        self.min_processing_time = float('inf')
        self.max_processing_time = 0

    def start(self):
        try:
            self.gps_thread.start()
            self.realsense_thread.start()
            self.processing_thread.start()
            logging.info("All threads started successfully.")
        except Exception as e:
            logging.error(f"Error starting threads: {str(e)}")
            self.stop()

    def stop(self):
        self.stop_event.set()
        for thread in [self.gps_thread, self.realsense_thread, self.processing_thread]:
            if thread.is_alive():
                thread.join()
        self.close()

    def gps_loop(self):
        try:
            gps_ser = serial.Serial(GPS_PORT, GPS_BAUD_RATE, timeout=1)
            logging.info("GPS connection established.")
        except serial.SerialException as e:
            logging.error(f"Unable to open GPS port: {str(e)}")
            self.stop_event.set()
            return

        while not self.stop_event.is_set():
            try:
                if gps_ser.in_waiting:
                    line = gps_ser.readline().decode('ascii', errors='ignore')
                    parsed_data = self.parse_nmea_data(line)
                    if parsed_data:
                        timestamp, lat, lon, heading = parsed_data
                        if heading is not None:
                            self.current_heading = heading
                        if lat is not None and lon is not None:
                            with self.gps_lock:
                                self.latest_gps = (timestamp, lat, lon, self.current_heading)
                            self.gps_file.write(f"{timestamp},{lat},{lon},{self.current_heading}\n")
                            self.gps_file.flush()
                            self.processing_queue.put(('gps', (timestamp, lat, lon, self.current_heading)))
            except serial.SerialException as e:
                logging.error(f"GPS read error: {str(e)}")
                self.stop_event.set()
                break
        
        gps_ser.close()

    def realsense_loop(self):
        try:
            self.pipeline = rs.pipeline()
            config = rs.config()
            config.enable_stream(rs.stream.depth, 848, 480, rs.format.z16, 30)
            config.enable_stream(rs.stream.color, 848, 480, rs.format.bgr8, 30)
            
            with self.pipeline_lock:
                self.pipeline.start(config)
                self.pipeline_started = True
            
            self.align = rs.align(rs.stream.color)
            
            logging.info("RealSense camera initialized successfully.")
        except rs.error as e:
            logging.error(f"Failed to initialize RealSense camera: {str(e)}")
            self.stop_event.set()
            return

        try:
            while not self.stop_event.is_set():
                frames = self.pipeline.wait_for_frames()
                aligned_frames = self.align.process(frames)
                
                depth_frame = aligned_frames.get_depth_frame()
                color_frame = aligned_frames.get_color_frame()
                
                if not depth_frame or not color_frame:
                    continue
                
                pc = rs.pointcloud()
                pc.map_to(color_frame)
                points = pc.calculate(depth_frame)
                
                vtx = np.asanyarray(points.get_vertices())
                pointcloud = np.array([(v[0], v[1], v[2]) for v in vtx])
                
                with self.pointcloud_lock:
                    self.latest_pointcloud = pointcloud
                                
        except rs.error as e:
            logging.error(f"RealSense camera error: {str(e)}")
        finally:
            self.stop_pipeline()

    def stop_pipeline(self):
        with self.pipeline_lock:
            if self.pipeline and self.pipeline_started:
                try:
                    self.pipeline.stop()
                    self.pipeline_started = False
                    logging.info("RealSense pipeline stopped successfully.")
                except Exception as e:
                    logging.error(f"Error stopping RealSense pipeline: {str(e)}")

    def processing_loop(self):
        while not self.stop_event.is_set():
            try:
                item = self.processing_queue.get(timeout=1)
                if item[0] == 'gps':
                    process_start_time = time.time()
                    timestamp, lat, lon, heading = item[1]
                    with self.pointcloud_lock:
                        if self.latest_pointcloud is not None:
                            np.save(self.pointcloud_file, {'timestamp': timestamp, 'points': self.latest_pointcloud})
                            self.pointcloud_file.flush()
                            processed_points = self.process_pointcloud(self.latest_pointcloud, (lat, lon, heading))
                            for point in processed_points:
                                self.processed_file.write(f"{timestamp},{point[0]},{point[1]},{point[2]}\n")
                            self.processed_file.flush()
                            process_end_time = time.time()
                            processing_time = process_end_time - process_start_time
                            total_time = process_end_time - timestamp
                            self.log_performance(timestamp, processing_time, total_time)
                            logging.info(f"Processed and saved data for timestamp {timestamp}")
                self.processing_queue.task_done()
            except queue.Empty:
                continue
            except Exception as e:
                logging.error(f"Error in processing loop: {str(e)}")
                if self.stop_event.is_set():
                    break

    def parse_nmea_data(self, data):
        data = data.strip().split(',')
        data_type = data[0][1:]

        if data_type == "GNGGA":
            try:
                timestamp = time.time() - self.start_time
                lat = float(data[2][:2]) + float(data[2][2:]) / 60
                if data[3] == 'S':
                    lat = -lat
                lon = float(data[4][:3]) + float(data[4][3:]) / 60
                if data[5] == 'W':
                    lon = -lon
                return timestamp, lat, lon, None
            except (ValueError, IndexError):
                logging.warning(f"Invalid GNGGA data: {data}")
                return None
        elif data_type == "GNRMC":
            try:
                timestamp = time.time() - self.start_time
                heading = float(data[8]) if data[8] else None
                return timestamp, None, None, heading
            except (ValueError, IndexError):
                logging.warning(f"Invalid GNRMC data: {data}")
                return None

        return None

    def process_pointcloud(self, pointcloud, gps_data):
        voxel_size = 0.05  # 5cm voxel size
        voxel_grid = {}
        for point in pointcloud:
            voxel_key = tuple(np.floor(point / voxel_size).astype(int))
            if voxel_key not in voxel_grid or point[2] < voxel_grid[voxel_key][2]:
                voxel_grid[voxel_key] = point
        downsampled_pointcloud = np.array(list(voxel_grid.values()))

        lat, lon, heading = gps_data
        easting, northing, _, _ = utm.from_latlon(lat, lon)

        R_tilt = Rotation.from_euler('x', SENSOR_TILT-90, degrees=True).as_matrix()
        R_orientation = Rotation.from_euler('z', SENSOR_ORIENTATION + heading, degrees=True).as_matrix()
        R = R_orientation @ R_tilt

        transformed_points = (R @ downsampled_pointcloud.T).T
        transformed_points[:, 0] += DISTANCE_FROM_GPS * np.sin(np.radians(ANGLE_FROM_GPS + heading))
        transformed_points[:, 1] += DISTANCE_FROM_GPS * np.cos(np.radians(ANGLE_FROM_GPS + heading))
        transformed_points[:, 2] += SENSOR_HEIGHT

        distances = np.linalg.norm(transformed_points, axis=1)
        transformed_points = transformed_points[distances <= MAX_DISTANCE_FROM_SENSOR]

        transformed_points[:, 0] += easting
        transformed_points[:, 1] += northing

        return transformed_points

    def log_performance(self, timestamp, processing_time, total_time):
        self.total_processing_time += processing_time
        self.total_processed_frames += 1
        self.min_processing_time = min(self.min_processing_time, processing_time)
        self.max_processing_time = max(self.max_processing_time, processing_time)
        
        avg_processing_time = self.total_processing_time / self.total_processed_frames
        
        performance_log = (f"Timestamp: {timestamp:.3f}, "
                           f"Processing Time: {processing_time:.3f}s, "
                           f"Total Time: {total_time:.3f}s, "
                           f"Avg Processing Time: {avg_processing_time:.3f}s, "
                           f"Min Processing Time: {self.min_processing_time:.3f}s, "
                           f"Max Processing Time: {self.max_processing_time:.3f}s\n")
        
        self.performance_file.write(performance_log)
        self.performance_file.flush()
        
        logging.debug(performance_log.strip())

    def close(self):
        logging.info("Closing connections and files...")
        self.stop_pipeline()
        self.gps_file.close()
        self.pointcloud_file.close()
        self.processed_file.close()
        self.performance_file.close()
        logging.info("All connections and files closed.")

def plot_data(data_folder):
    logging.info("Plotting data...")
    try:
        processed_data = np.loadtxt(os.path.join(data_folder, 'processed_data.txt'), delimiter=',')

        if processed_data.size == 0 or processed_data.ndim == 1:
            logging.warning("Insufficient data for plotting.")
            return

        timestamps, x, y, z = processed_data[:, 0], processed_data[:, 1], processed_data[:, 2], processed_data[:, 3]

        x_rel = x - np.min(x)
        y_rel = y - np.min(y)

        fig_3d = go.Figure(data=[go.Scatter3d(
            x=x_rel,
            y=y_rel,
            z=z,
            mode='markers',
            marker=dict(
                size=3,
                color=z,
                colorscale='Viridis',
                opacity=1,
                colorbar=dict(title='Elevation (m)')
            )
        )])

        fig_3d.update_layout(
            scene=dict(
                aspectmode='data',
                xaxis_title='Relative Easting (m)',
                yaxis_title='Relative Northing (m)',
                zaxis_title='Elevation (m)'
            ),
            title='3D Point Cloud'
        )

        fig_3d.write_html(os.path.join(data_folder, '3d_pointcloud.html'))

        fig_2d = go.Figure(data=[go.Scatter(
            x=x_rel,
            y=y_rel,
            mode='markers',
            marker=dict(
                size=3,
                color=z,
                colorscale='Viridis',
                opacity=1,
                colorbar=dict(title='Elevation (m)')
            )
        )])

        fig_2d.update_layout(
            xaxis_title='Relative Easting (m)',
            yaxis_title='Relative Northing (m)',
            title='2D Point Cloud Projection'
        )

        fig_2d.write_html(os.path.join(data_folder, '2d_pointcloud.html'))

        logging.info(f"Plots saved in {data_folder}")
    except Exception as e:
        logging.error(f"An error occurred while plotting: {str(e)}")

def main():
    while True:
        key = input("Enter 'R' to record new data, 'P' to plot existing data, or 'Q' to quit: ").upper()
        
        if key == 'R':
            recorder = DataRecorder()
            try:
                recorder.start()
                input("Press Enter to stop recording...\n")
            except Exception as e:
                logging.error(f"Error during recording: {str(e)}")
            finally:
                recorder.stop()
            
            if os.path.exists(recorder.session_folder) and os.path.getsize(os.path.join(recorder.session_folder, 'processed_data.txt')) > 0:
                plot_data(recorder.session_folder)
            else:
                logging.warning("No data was recorded or the data file is empty. Unable to generate plots.")

        elif key == 'P':
            folder = input("Enter the path to the data folder: ")
            if os.path.exists(folder):
                plot_data(folder)
            else:
                logging.error("Invalid folder path.")

        elif key == 'Q':
            break
        else:
            logging.warning("Invalid choice. Please try again.")

if __name__ == "__main__":
    main()

File: v2/realsense/cpp/main.cpp
#define _WIN32_WINNT 0x0601 // Target Windows 7 or later
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <ctime>
#include <thread>
#include <queue>
#include <mutex>
#include <condition_variable>
#include <filesystem>
#include <atomic>
#include <chrono>
#include <cmath>
#include <algorithm>
#include <limits>
#include <vector>
#include <boost/asio.hpp>
#include <librealsense2/rs.hpp>
#include <Eigen/Dense>
#include <ogr_geometry.h>
#include <stdlib.h>

using namespace std::chrono;
namespace fs = std::filesystem;
using boost::asio::ip::tcp;

const std::string GPS_PORT = "COM4";
const int GPS_BAUD_RATE = 57600;
const double SENSOR_HEIGHT = 0.973; // Meters
const double SENSOR_TILT = 30; // Degrees
const double ANGLE_FROM_GPS = 270; // Degrees
const double DISTANCE_FROM_GPS = 0.4; // Meters
const double SENSOR_ORIENTATION = 2; // Degrees
const double MAX_DISTANCE_FROM_SENSOR = 4; // Meters
const double DOWNSCALE_FACTOR = 1;

class DataRecorder
{
public:
    DataRecorder() : start_time(system_clock::now()),
                     session_folder("data_" + currentDateTime()),
                     total_processing_time(0),
                     total_processed_frames(0),
                     min_processing_time(std::numeric_limits<double>::infinity()),
                     max_processing_time(0),
                     stop_event(false),
                     align_to_color(RS2_STREAM_COLOR)
    {
        fs::create_directory(session_folder);
        gps_file.open(session_folder + "/gps_data.txt");
        pointcloud_file.open(session_folder + "/pointcloud_data.npy", std::ios::binary);
        processed_file.open(session_folder + "/processed_data.txt");
        performance_file.open(session_folder + "/performance_data.txt");

        gps_thread = std::thread(&DataRecorder::gps_loop, this);
        realsense_thread = std::thread(&DataRecorder::realsense_loop, this);
        processing_thread = std::thread(&DataRecorder::processing_loop, this);
    }

    ~DataRecorder()
    {
        stop();
    }

    void start()
    {
        try
        {
            gps_thread.detach();
            realsense_thread.detach();
            processing_thread.detach();
            std::cout << "All threads started successfully." << std::endl;
        }
        catch (const std::exception &e)
        {
            std::cerr << "Error starting threads: " << e.what() << std::endl;
            stop();
        }
    }

    void stop()
    {
        stop_event = true;
        if (gps_thread.joinable())
            gps_thread.join();
        if (realsense_thread.joinable())
            realsense_thread.join();
        if (processing_thread.joinable())
            processing_thread.join();
        close();
    }

private:
    system_clock::time_point start_time;
    std::string session_folder;
    std::ofstream gps_file;
    std::ofstream pointcloud_file;
    std::ofstream processed_file;
    std::ofstream performance_file;
    std::vector<Eigen::Vector3d> latest_pointcloud;
    std::tuple<double, double, double, double> latest_gps;
    double current_heading;
    std::mutex pointcloud_lock;
    std::mutex gps_lock;
    std::queue<std::tuple<std::string, std::tuple<double, double, double, double>>> processing_queue;
    std::thread gps_thread;
    std::thread realsense_thread;
    std::thread processing_thread;
    rs2::pipeline pipeline;
    rs2::align align_to_color;
    std::atomic<bool> pipeline_started;
    std::mutex pipeline_lock;
    std::atomic<bool> stop_event;

    double total_processing_time;
    int total_processed_frames;
    double min_processing_time;
    double max_processing_time;

    static std::string currentDateTime()
    {
        auto now = std::chrono::system_clock::now();
        auto in_time_t = std::chrono::system_clock::to_time_t(now);
        struct tm timeinfo;
        localtime_s(&timeinfo, &in_time_t);
        std::stringstream ss;
        ss << std::put_time(&timeinfo, "%Y%m%d_%H%M%S");
        return ss.str();
    }

    void gps_loop()
    {
        try
        {
            boost::asio::io_service io_service;
            boost::asio::serial_port serial(io_service, GPS_PORT);
            serial.set_option(boost::asio::serial_port_base::baud_rate(GPS_BAUD_RATE));
            std::cout << "GPS connection established." << std::endl;

            while (!stop_event)
            {
                char c;
                std::string line;
                while (boost::asio::read(serial, boost::asio::buffer(&c, 1)) > 0 && c != '\n')
                {
                    line += c;
                }

                auto parsed_data = parse_nmea_data(line);
                if (std::get<0>(parsed_data) != 0)
                {
                    double timestamp = std::get<0>(parsed_data);
                    double lat = std::get<1>(parsed_data);
                    double lon = std::get<2>(parsed_data);
                    double heading = std::get<3>(parsed_data);
                    if (heading != 0)
                    {
                        current_heading = heading;
                    }
                    if (lat != 0 && lon != 0)
                    {
                        std::lock_guard<std::mutex> lock(gps_lock);
                        latest_gps = std::make_tuple(timestamp, lat, lon, current_heading);
                        gps_file << std::setprecision(15) << timestamp << "," << lat << "," << lon << "," << heading << std::endl;
                        gps_file.flush();
                        processing_queue.push(std::make_tuple("gps", latest_gps));
                    }
                }
            }
        }
        catch (const std::exception &e)
        {
            std::cerr << "GPS read error: " << e.what() << std::endl;
            stop_event = true;
        }
    }

    void realsense_loop()
    {
        try
        {
            rs2::config cfg;
            cfg.enable_stream(RS2_STREAM_DEPTH, 848, 480, RS2_FORMAT_Z16, 30);
            cfg.enable_stream(RS2_STREAM_COLOR, 848, 480, RS2_FORMAT_BGR8, 30);
            pipeline.start(cfg);
            pipeline_started = true;
            align_to_color = rs2::align(RS2_STREAM_COLOR);
            std::cout << "RealSense camera initialized successfully." << std::endl;

            while (!stop_event)
            {
                rs2::frameset frames = pipeline.wait_for_frames();
                frames = align_to_color.process(frames);

                rs2::frame depth_frame = frames.get_depth_frame();
                rs2::frame color_frame = frames.get_color_frame();

                if (!depth_frame || !color_frame)
                    continue;

                rs2::pointcloud pc;
                pc.map_to(color_frame);
                rs2::points points = pc.calculate(depth_frame);

                auto vertices = points.get_vertices();
                std::vector<Eigen::Vector3d> pointcloud;
                for (int i = 0; i < points.size(); ++i)
                {
                    pointcloud.push_back(Eigen::Vector3d(vertices[i].x, vertices[i].y, vertices[i].z));
                }

                std::lock_guard<std::mutex> lock(pointcloud_lock);
                latest_pointcloud = pointcloud;

                std::string filename = session_folder + "/pointcloud_" + std::to_string(std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()) + ".ply";
                points.export_to_ply(filename, color_frame);
            }
        }
        catch (const rs2::error &e)
        {
            std::cerr << "RealSense camera error: " << e.what() << std::endl;
        }
        stop_pipeline();
    }

    void stop_pipeline()
    {
        std::lock_guard<std::mutex> lock(pipeline_lock);
        if (pipeline_started)
        {
            try
            {
                pipeline.stop();
                pipeline_started = false;
                std::cout << "RealSense pipeline stopped successfully." << std::endl;
            }
            catch (const std::exception &e)
            {
                std::cerr << "Error stopping RealSense pipeline: " << e.what() << std::endl;
            }
        }
    }

    void processing_loop()
    {
        while (!stop_event)
        {
            try
            {
                if (!processing_queue.empty())
                {
                    auto item = processing_queue.front();
                    processing_queue.pop();
                    if (std::get<0>(item) == "gps")
                    {
                        auto process_start_time = system_clock::now();
                        double timestamp = std::get<0>(std::get<1>(item));
                        double lat = std::get<1>(std::get<1>(item));
                        double lon = std::get<2>(std::get<1>(item));
                        double heading = std::get<3>(std::get<1>(item));
                        std::lock_guard<std::mutex> lock(pointcloud_lock);
                        if (!latest_pointcloud.empty())
                        {
                            auto processed_pointcloud = process_pointcloud(latest_pointcloud, std::make_tuple(lat, lon, heading));

                            for (const auto &point : processed_pointcloud)
                            {
                                processed_file << std::setprecision(15) << point[0] << "," << point[1] << "," << point[2] << std::endl;
                            }
                            processed_file.flush();

                            auto process_end_time = system_clock::now();
                            auto processing_time = duration_cast<microseconds>(process_end_time - process_start_time).count();
                            auto total_time = duration_cast<microseconds>(process_end_time - start_time).count();
                            log_performance(timestamp, processing_time / 1000000.0, total_time / 1000000.0);
                            std::cout << "Processed and saved data for timestamp " << timestamp << std::endl;
                        }
                    }
                }
            }
            catch (const std::exception &e)
            {
                std::cerr << "Error in processing loop: " << e.what() << std::endl;
                if (stop_event)
                    break;
            }
        }
    }

    std::tuple<double, double, double, double> parse_nmea_data(const std::string &data)
    {
        std::stringstream ss(data);
        std::string item;
        std::vector<std::string> tokens;
        while (std::getline(ss, item, ','))
        {
            tokens.push_back(item);
        }
        std::string data_type = tokens[0].substr(1);

        if (data_type == "GNGGA" && tokens.size() >= 6)
        {
            try
            {
                double timestamp = duration_cast<milliseconds>(system_clock::now() - start_time).count() / 1000.0;
                double lat = std::stod(tokens[2].substr(0, 2)) + std::stod(tokens[2].substr(2)) / 60.0;
                if (tokens[3] == "S")
                    lat = -lat;
                double lon = std::stod(tokens[4].substr(0, 3)) + std::stod(tokens[4].substr(3)) / 60.0;
                if (tokens[5] == "W")
                    lon = -lon;

                if (lat < -90 || lat > 90 || lon < -180 || lon > 180)
                {
                    std::cerr << "Invalid GPS coordinates: " << lat << ", " << lon << std::endl;
                    return std::make_tuple(0, 0, 0, 0);
                }

                return std::make_tuple(timestamp, lat, lon, 0);
            }
            catch (const std::exception &e)
            {
                std::cerr << "Error parsing GNGGA data: " << e.what() << std::endl;
                return std::make_tuple(0, 0, 0, 0);
            }
        }
        else if (data_type == "GNRMC" && tokens.size() >= 9)
        {
            try
            {
                double timestamp = duration_cast<milliseconds>(system_clock::now() - start_time).count() / 1000.0;
                double heading = tokens[8].empty() ? 0 : std::stod(tokens[8]);
                return std::make_tuple(timestamp, 0, 0, heading);
            }
            catch (const std::exception &e)
            {
                std::cerr << "Error parsing GNRMC data: " << e.what() << std::endl;
                return std::make_tuple(0, 0, 0, 0);
            }
        }
        return std::make_tuple(0, 0, 0, 0);
    }

    std::vector<Eigen::Vector3d> process_pointcloud(const std::vector<Eigen::Vector3d> &pointcloud, const std::tuple<double, double, double> &gps_data)
    {
        const double voxel_size = 0.05; // 5cm voxel size
        std::map<std::tuple<int, int, int>, Eigen::Vector3d> voxel_grid;
        for (const auto &point : pointcloud)
        {
            auto voxel_key = std::make_tuple(
                static_cast<int>(std::floor(point[0] / voxel_size)),
                static_cast<int>(std::floor(point[1] / voxel_size)),
                static_cast<int>(std::floor(point[2] / voxel_size)));
            if (voxel_grid.find(voxel_key) == voxel_grid.end() || point[2] < voxel_grid[voxel_key][2])
            {
                voxel_grid[voxel_key] = point;
            }
        }
        std::vector<Eigen::Vector3d> downsampled_pointcloud;
        for (const auto &voxel : voxel_grid)
        {
            downsampled_pointcloud.push_back(voxel.second);
        }

        double lat = std::get<0>(gps_data);
        double lon = std::get<1>(gps_data);
        double heading = std::get<2>(gps_data);
        std::cout << "GPS data: " << std::setprecision(15) << lat << ", " << std::setprecision(15) << lon << ", " << heading << std::endl;

        if (lat == 0 && lon == 0)
        {
            std::cerr << "Invalid GPS data. Skipping point cloud processing." << std::endl;
            return std::vector<Eigen::Vector3d>();
        }

        double easting = 0, northing = 0;
        OGRSpatialReference sr;
        sr.SetWellKnownGeogCS("WGS84");
        OGRSpatialReference utm;
        int zone = static_cast<int>((lon + 180) / 6) + 1;
        utm.SetUTM(zone, lat >= 0);
        OGRCoordinateTransformation *transform = OGRCreateCoordinateTransformation(&sr, &utm);
        if (transform && transform->Transform(1, &lat, &lon))
        {
            easting = lon;
            northing = lat;
        }
        else
        {
            std::cerr << "Transformation error: " << CPLGetLastErrorMsg() << std::endl;
            easting = lon;
            northing = lat;
        }

        OGRCoordinateTransformation::DestroyCT(transform);

        Eigen::Matrix3d R_tilt = Eigen::AngleAxisd(-(SENSOR_TILT+90) * M_PI / 180, Eigen::Vector3d::UnitX()).toRotationMatrix();
        Eigen::Matrix3d R_orientation = Eigen::AngleAxisd((SENSOR_ORIENTATION + heading) * M_PI / 180, Eigen::Vector3d::UnitZ()).toRotationMatrix();
        Eigen::Matrix3d R = R_orientation * R_tilt;

        std::vector<Eigen::Vector3d> transformed_points;
        for (const auto &point : downsampled_pointcloud)
        {
            Eigen::Vector3d transformed_point = R * point;
            transformed_point[0] += DISTANCE_FROM_GPS * std::sin((ANGLE_FROM_GPS + heading) * M_PI / 180);
            transformed_point[1] += DISTANCE_FROM_GPS * std::cos((ANGLE_FROM_GPS + heading) * M_PI / 180);
            transformed_point[2] += SENSOR_HEIGHT;
            double distance = transformed_point.norm();
            if (distance <= MAX_DISTANCE_FROM_SENSOR)
            {
                transformed_point[0] += easting;
                transformed_point[1] += northing;
                if (std::isfinite(transformed_point[0]) && std::isfinite(transformed_point[1]) && std::isfinite(transformed_point[2]))
                {
                    transformed_points.push_back(transformed_point);
                }
                else
                {
                    std::cerr << "Invalid transformed point: " << transformed_point.transpose() << std::endl;
                }
            }
        }
        return transformed_points;
    }

    void log_performance(double timestamp, double processing_time, double total_time)
    {
        total_processing_time += processing_time;
        total_processed_frames++;
        min_processing_time = std::min(min_processing_time, processing_time);
        max_processing_time = std::max(max_processing_time, processing_time);
        double avg_processing_time = total_processing_time / total_processed_frames;

        std::stringstream ss;
        ss << "Timestamp: " << timestamp << ", "
           << "Processing Time: " << processing_time << "s, "
           << "Total Time: " << total_time << "s, "
           << "Avg Processing Time: " << avg_processing_time << "s, "
           << "Min Processing Time: " << min_processing_time << "s, "
           << "Max Processing Time: " << max_processing_time << "s" << std::endl;

        performance_file << ss.str();
        performance_file.flush();
        std::cout << ss.str();
    }

    void close()
    {
        std::cout << "Closing connections and files..." << std::endl;
        stop_pipeline();
        if (gps_file.is_open())
            gps_file.close();
        if (pointcloud_file.is_open())
            pointcloud_file.close();
        if (processed_file.is_open())
            processed_file.close();
        if (performance_file.is_open())
            performance_file.close();
        std::cout << "All connections and files closed." << std::endl;
    }
};

void main_loop()
{
    while (true)
    {
        std::cout << "Enter 'R' to record new data, 'Q' to quit: ";
        char key;
        std::cin >> key;
        key = std::toupper(key);

        if (key == 'R')
        {
            DataRecorder recorder;
            try
            {
                recorder.start();
                std::cout << "Press Enter to stop recording...\n";
                std::cin.ignore();
                std::cin.get();
            }
            catch (const std::exception &e)
            {
                std::cerr << "Error during recording: " << e.what() << std::endl;
            }
            recorder.stop();
        }
        else if (key == 'Q')
        {
            break;
        }
        else
        {
            std::cerr << "Invalid choice. Please try again." << std::endl;
        }
    }
}

int main()
{
    const char *proj_path = "C:\\OSGeo4W\\share\\proj";
    _putenv_s("PROJ_LIB", proj_path);
    main_loop();
    return 0;
}
